api/objectstore.go

```diff
@@ -0,0 +1,193 @@
+package api
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/longhorn/longhorn-manager/datastore"
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+	"github.com/longhorn/longhorn-manager/util"
+	"github.com/pkg/errors"
+	"github.com/rancher/go-rancher/api"
+	"github.com/rancher/go-rancher/client"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func (s *Server) ObjectStoreList(rw http.ResponseWriter, req *http.Request) (err error) {
+	apiContext := api.GetApiContext(req)
+	col, err := s.objectStoreList(apiContext)
+	if err != nil {
+		return err
+	}
+	apiContext.Write(col)
+	return nil
+}
+
+func (s *Server) objectStoreList(apiContext *api.ApiContext) (*client.GenericCollection, error) {
+	list, err := s.m.ListObjectStoresSorted()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list object stores")
+	}
+
+	data := []interface{}{}
+
+	for _, store := range list {
+		vol, err := s.m.Get(fmt.Sprintf("pv-%v", store.Name))
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get volume for %v", store.Name)
+		}
+		data = append(data, toObjectStoreResource(store, vol.Spec.Size, vol.Status.ActualSize))
+	}
+
+	return &client.GenericCollection{
+		Data: data,
+		Collection: client.Collection{
+			ResourceType: "objectStore",
+		},
+	}, nil
+}
+
+func (s *Server) ObjectStoreGet(rw http.ResponseWriter, req *http.Request) (err error) {
+	apiContext := api.GetApiContext(req)
+	resp := &client.GenericCollection{}
+	apiContext.Write(resp)
+	return nil
+}
+
+func (s *Server) ObjectStoreCreate(rw http.ResponseWriter, req *http.Request) (err error) {
+	apiContext := api.GetApiContext(req)
+
+	var input ObjectStoreInput
+	if err := apiContext.Read(&input); err != nil {
+		return err
+	}
+
+	size := resource.MustParse(input.Size)
+
+	// find a new name for the secret and create a new secret to seed the
+	// credentials of the object store.
+	var secretName string = input.Name
+	for {
+		_, err := s.m.GetSecret(s.m.GetLonghornNamespace(), secretName)
+		if err != nil && datastore.ErrorIsNotFound(err) {
+			break
+		} else if err != nil {
+			return errors.Wrapf(err, "API error while searching for secret %v", secretName)
+		} else {
+			secretName = fmt.Sprintf("%v-%v", input.Name, util.RandomString(6))
+		}
+	}
```

Comment:
The current flow might leave a bunch of orphan secrete after repeatedly create/delete a ObjectStore with the same name because we keep generating new random secrete name each time. 

I am thinking about this design:
1. On the UI page when user creates a new ObjectStore we allow the user to choose between 2 options:
    1. Create a new secret with the specified AccessKey/SecretKey.
    2. Users already prepared a secret by themselves. Don't need to generate AccessKey/SecretKey. Users can enter the already existing secret ref 
2. On the backend,
    1. When user choose option 1 above, the API will attempt to create a new secret with the same name as the ObjectStore name. If the secret already exist, backend returns error. User would need to cleanup the existing secret OR reuse the secret by using option 2.
    2. When the user choose option 2, the backend search for the secret name and use it. If the specified secret ref doesn't exist, backend returns error. 
3. This design will work for the kubectl flow also. User must deploy a secret first and deploy the ObjectStore yaml that reference the existing secret.
4. For the backup, we don't backup the secret because the AccessKey/SecretKey are already burned into the volume's data. 
5. For the restore flow, 
    1. UI doesn't allow user to specify secret ref or create a new one because the data is already burned into the volume
    2. When backend received a request with `FromBackup` field, it create a new ObjectStore with an empty secret ref
    3. Similarly when user restore an ObjectStore using kubectl, they create a new ObjectStore yaml with empty Spec.Credentials and non-empty Spec.Storage.FromBackup

Reply:
In the current desing, if the object store is created from the UI, the secret is created by the manager API server. Why don't we just add the owner reference to the secret and let it be deleted when the object store is deleted?
This would still work with kubectl, but the user would need to manage creation/deletion of the secret, which makes sense because the secret is a separate K8s object.
This would avoid the littering problem and never show an error to the user, since there is nothing that can be done wrong.

I'm not a fan of letting the user choose things in the UI, this makes the dialogues more complicated and needs explanation in documentation.

Reply:
I see your point about `not a fan of letting the user choose things in the UI, this makes the dialogues more complicated and needs explanation in documentation.`

Incorporating your feedback, how do you think about this updated design below?

1. On the UI page, when user creates a new ObjectStore users always have to specify AccessKey/SecretKey.
2. On the backend,
    1. The API will attempt to create a new secret with the same name as the ObjectStore name. We also append a add special label `longhorn.io/object-store: <object-store-name>` and `longhorn.io/managed-by: longhorn-manager` to the secret 
    2. If the secret already exist, Check of the secrete has the label `longhorn.io/object-store: <object-store-name>` and `longhorn.io/managed-by: longhorn-manager`
        1. If yes, adopt the secret and modify the AccessKey/SecretKey to be the one user specified
        2. If no, return error. This secret is not managed by longhorn so user would need to cleanup it
5. For kubectl flow also., user must deploy a secret first and deploy the ObjectStore yaml that reference the existing secret.
6. For the backup, we don't backup the secret because the AccessKey/SecretKey are already burned into the volume's data. 
7. For the restore flow, 
    1. UI doesn't allow user to specify secret ref or create a new one because the data is already burned into the volume
    4. When backend received a request with `FromBackup` field, it creates a new ObjectStore with an empty `Spec.Credentials`
    5. Similarly when users restore an ObjectStore using kubectl, they create a new ObjectStore yaml with empty `Spec.Credentials` and non-empty `Spec.Storage.FromBackup`
1. When the users delete an ObjectStore:
    1. ObjectStore controller will make sure to list all secrets with the label `longhorn.io/object-store: <object-store-name>` and `longhorn.io/managed-by: longhorn-manager`  and delete them before removing the finalizer for the ObjectStore

Reply:
This seems like a sensible design, I'll implement it